protocol/directory.go

```diff
@@ -91,14 +109,77 @@ func (d *ConiksDirectory) Register(req *RegistrationRequest) (
 		// insert new data to the directory on-the-fly
 		tb, err := d.pad.TB(req.Username, req.Key)
 		if err != nil {
-			return nil, ErrorDirectory
+			return NewErrorResponse(ErrorDirectory), ErrorDirectory
 		}
 		d.tbs[req.Username] = tb
 		return NewRegistrationProof(ap, d.LatestSTR(), tb, Success)
 	} else {
 		if err = d.pad.Set(req.Username, req.Key); err != nil {
-			return nil, ErrorDirectory
+			return NewErrorResponse(ErrorDirectory), ErrorDirectory
 		}
 		return NewRegistrationProof(ap, d.LatestSTR(), nil, Success)
 	}
 }
+
+func (d *ConiksDirectory) KeyLookup(req *KeyLookupRequest) (
+	Response, ErrorCode) {
+	ap, err := d.pad.Lookup(req.Username)
+	if err != nil {
+		return NewErrorResponse(ErrorDirectory), ErrorDirectory
+	}
+
+	if !bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		// if not found in the tree, do lookup in tb array
+		if d.useTBs {
+			if tb := d.tbs[req.Username]; tb != nil {
+				return NewKeyLookupProof(ap, d.LatestSTR(), tb, Success)
+			}
+		}
+		return NewKeyLookupProof(ap, d.LatestSTR(), nil, ErrorNameNotFound)
+	}
+	return NewKeyLookupProof(ap, d.LatestSTR(), nil, Success)
```

Comment:
Just a suggestion.  Little clearer like,

```
if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
  return NewKeyLookupInEpochProof(ap, strs, Success)
}
// if not found in the tree, do lookup in tb array
if d.useTBs {
  if tb := d.tbs[req.Username]; tb != nil {
    return NewKeyLookupProof(ap, d.LatestSTR(), tb, Success)
  }
}
return NewKeyLookupProof(ap, d.LatestSTR(), nil, ErrorNameNotFound)
```

You'd want to switch up the end of `KeyLookupInEpoch` to be the same though.